File: cli/internal_testsuites/golang/testsuite/basic_datastore_and_api_test/basic_datastore_and_api_test.go
package basic_datastore_and_api_test

import (
	"context"
	"github.com/kurtosis-tech/example-api-server/api/golang/example_api_server_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis-cli/golang_internal_testsuite/test_helpers"
	"github.com/kurtosis-tech/kurtosis-core-api-lib/api/golang/lib/services"
	"github.com/sirupsen/logrus"
	"github.com/stretchr/testify/require"
	"testing"
)

const (
	testName = "basic-datastore-and-api"
	isPartitioningEnabled = false

	datastoreServiceId services.ServiceID = "datastore"
	apiServiceId       services.ServiceID = "api"

	testPersonId     = "23"
	testNumBooksRead = uint32(3)
)

func TestBasicDatastoreAndAPITest(t *testing.T) {
	ctx := context.Background()

	// ------------------------------------- ENGINE SETUP ----------------------------------------------
	enclaveCtx, destroyEnclaveFunc, _, err := test_helpers.CreateEnclave(t, ctx, testName, isPartitioningEnabled)
	require.NoError(t, err, "An error occurred creating an enclave")
	defer destroyEnclaveFunc()


	// ------------------------------------- TEST SETUP ----------------------------------------------
	// TODO replace with datastore launcher inside the lib
	logrus.Infof("Adding datastore service...")
	datastoreServiceCtx, _, datastoreClientCloseFunc, err := test_helpers.AddDatastoreService(ctx, datastoreServiceId, enclaveCtx)
	require.NoError(t, err, "An error occurred adding the datastore service to the enclave")
	defer datastoreClientCloseFunc()
	logrus.Infof("Added datastore service")

	logrus.Infof("Adding API service...")
	_, apiClient, apiClientCloseFunc, err := test_helpers.AddAPIService(ctx, apiServiceId, enclaveCtx, datastoreServiceCtx.GetPrivateIPAddress())
	require.NoError(t, err, "An error occurred adding the API service to the enclave")
	defer apiClientCloseFunc()
	logrus.Infof("Added API service")

	// ------------------------------------- TEST RUN ----------------------------------------------
	logrus.Infof("Verifying that person with test ID '%v' doesn't already exist...", testPersonId)
	getPersonArgs := &example_api_server_rpc_api_bindings.GetPersonArgs{
		PersonId: testPersonId,
	}
	 _, err = apiClient.GetPerson(ctx, getPersonArgs)
	require.Error(t, err, "Expected an error trying to get a person who doesn't exist yet, but didn't receive one")
	logrus.Infof("Verified that test person doesn't already exist")

	logrus.Infof("Adding test person with ID '%v'...", testPersonId)
	addPersonArgs := &example_api_server_rpc_api_bindings.AddPersonArgs{
		PersonId: testPersonId,
	}
	 _, err = apiClient.AddPerson(ctx, addPersonArgs)
	require.NoError(t, err, "An error occurred adding test person with ID '%v'", testPersonId)
	logrus.Info("Test person added")

	logrus.Infof("Incrementing test person's number of books read by %v...", testNumBooksRead)
	for i := uint32(0); i < testNumBooksRead; i++ {
		incrementBooksReadArgs := &example_api_server_rpc_api_bindings.IncrementBooksReadArgs{
			PersonId: testPersonId,
		}
		_, err = apiClient.IncrementBooksRead(ctx, incrementBooksReadArgs)
		require.NoError(t, err, "An error occurred incrementing the number of books read")
	}
	logrus.Info("Incremented number of books read")

	logrus.Info("Retrieving test person to verify number of books read...")
	getPersonResponse, err := apiClient.GetPerson(ctx, getPersonArgs)
	require.NoError(t, err, "An error occurred getting the test person to verify the number of books read")
	logrus.Info("Retrieved test person")

	require.Equal(
		t,
		testNumBooksRead,
		getPersonResponse.GetBooksRead(),
		"Expected number of book read '%v' != actual number of books read '%v'",
		testNumBooksRead,
		getPersonResponse.GetBooksRead(),
	)
}

File: internal_testsuites/typescript/src/testsuite/basic_datastore_and_api_test/basic_datastore_and_api.test.ts
import { ServiceContext } from "kurtosis-sdk"
import * as apiServerApi from "example-api-server-api-lib";
import { Result, ok, err } from "neverthrow"
import log from "loglevel"
import * as grpc from "@grpc/grpc-js"
import * as google_protobuf_empty_pb from "google-protobuf/google/protobuf/empty_pb";

import { createEnclave } from "../../test_helpers/enclave_setup";
import { addAPIService, addDatastoreService } from "../../test_helpers/test_helpers";

const TEST_NAME = "basic-datastore-and-api";
const IS_PARTITIONING_ENABLED = false;
const DATASTORE_SERVICE_NAME = "datastore";
const API_SERVICE_NAME = "api";
const TEST_PERSON_ID = "23";
const TEST_NUM_BOOKS_READ = 3;

jest.setTimeout(180000)

test("Test basic data store and API", async () => {
    // ------------------------------------- ENGINE SETUP ----------------------------------------------
    const createEnclaveResult = await createEnclave(TEST_NAME, IS_PARTITIONING_ENABLED)

    if(createEnclaveResult.isErr()) { throw createEnclaveResult.error }

    const { enclaveContext, stopEnclaveFunction } = createEnclaveResult.value

    try {
        // ------------------------------------- TEST SETUP ----------------------------------------------

        log.info("Adding datastore service...")

        const addDatastoreServiceResult = await addDatastoreService(DATASTORE_SERVICE_NAME, enclaveContext)

        if(addDatastoreServiceResult.isErr()) { throw addDatastoreServiceResult.error }

        const { 
            serviceContext: datastoreServiceContext, 
            clientCloseFunction:datastoreClientCloseFunction 
        } = addDatastoreServiceResult.value

        log.info("Added datastore service")

        try {

            log.info("Adding API service...")
            const apiClientServiceResult: Result<{
                serviceContext: ServiceContext;
                client: apiServerApi.ExampleAPIServerServiceClientNode;
                clientCloseFunction: () => void;
            }, Error> = await addAPIService(API_SERVICE_NAME, enclaveContext, datastoreServiceContext.getPrivateIPAddress())
            
            if(apiClientServiceResult.isErr()){ throw apiClientServiceResult.error }

            const { 
                client: apiClient, 
                clientCloseFunction: apiClientCloseFunction  
            } = apiClientServiceResult.value

                log.info("Added API service")
            
            try {
                // ------------------------------------- TEST RUN ----------------------------------------------
                log.info(`Verifying that person with test ID ${TEST_PERSON_ID} doesn't already exist...`);
                
                const getPersonArgs = new apiServerApi.GetPersonArgs()
                getPersonArgs.setPersonId(TEST_PERSON_ID)

                const getPersonResultPromise: Promise<Result<apiServerApi.GetPersonResponse, Error>> = new Promise((resolve, _unusedReject) => {
                        apiClient.getPerson(getPersonArgs, (error: grpc.ServiceError | null, response?: apiServerApi.GetPersonResponse) => {
                            if (error === null) {
                                if (!response) {
                                    resolve(err(new Error("No error was encountered but the response was still falsy; this should never happen")));
                                } else {
                                    resolve(ok(response!));
                                }
                            } else {
                                resolve(err(error));
                            }
                        })
                })

                const getPersonResult = await getPersonResultPromise;
                if(getPersonResult.isOk()) { 
                    throw new Error("Expected an error trying to get a person who doesn't exist yet, but didn't receive one")
                }
                log.info("Verified that test person doesn't already exist")
                
                log.info(`Adding test person with ID ${TEST_PERSON_ID}...`)

                const addPersonArgs = new apiServerApi.AddPersonArgs()
                addPersonArgs.setPersonId(TEST_PERSON_ID)

                const addPersonResultPromise: Promise<Result<google_protobuf_empty_pb.Empty, Error>> = new Promise((resolve, _unusedReject) => {
                    apiClient.addPerson(addPersonArgs, (error: grpc.ServiceError | null, response?: google_protobuf_empty_pb.Empty) => {
                        if (error === null) {
                            if (!response) {
                                resolve(err(new Error("No error was encountered but the response was still falsy; this should never happen")));
                            } else {
                                resolve(ok(response!));
                            }
                        } else {
                            resolve(err(error));
                        }
                    })
                })

                const addPersonResult = await addPersonResultPromise;
                if(addPersonResult.isErr()) {
                    log.error(`An error occurred adding test person with ID ${TEST_PERSON_ID}`)
                    throw addPersonResult.error;
                }
                log.info("Test person added")
                
                log.info(`Incrementing test person's number of books read by ${TEST_NUM_BOOKS_READ}...`)
                
                const incrementBooksReadArgs = new apiServerApi.IncrementBooksReadArgs()
                incrementBooksReadArgs.setPersonId(TEST_PERSON_ID)
            
                for (let i = 0; i < TEST_NUM_BOOKS_READ; i++) {
                    const incrementBooksReadResultPromise: Promise<Result<google_protobuf_empty_pb.Empty, Error>> = new Promise((resolve, _unusedReject) => {
                        apiClient.incrementBooksRead(incrementBooksReadArgs, (error: grpc.ServiceError | null, response?: google_protobuf_empty_pb.Empty) => {
                            if (error === null) {
                                if (!response) {
                                    resolve(err(new Error("No error was encountered but the response was still falsy; this should never happen")));
                                } else {
                                    resolve(ok(response!));
                                }
                            } else {
                                resolve(err(error));
                            }
                        })
                    })
                    const incrementBooksReadResult = await incrementBooksReadResultPromise;
                    if(incrementBooksReadResult.isErr()) { 
                        log.error("An error occurred incrementing the number of books read")
                        throw incrementBooksReadResult.error;
                    }
                }
                
                log.info("Incremented number of books read")
                
                log.info("Retrieving test person to verify number of books read...")
                
                const getPersonAfterUpdatingResultPromise: Promise<Result<apiServerApi.GetPersonResponse, Error>> = new Promise((resolve, _unusedReject) => {
                    apiClient.getPerson(getPersonArgs, (error: grpc.ServiceError | null, response?: apiServerApi.GetPersonResponse) => {
                        if (error === null) {
                            if (!response) {
                                resolve(err(new Error("No error was encountered but the response was still falsy; this should never happen")));
                            } else {
                                resolve(ok(response!));
                            }
                        } else {
                            resolve(err(error));
                        }
                    })
                })
                
                const  getPersonAfterUpdatingResult = await getPersonAfterUpdatingResultPromise
                if(getPersonAfterUpdatingResult.isErr()){
                    log.error("An error occurred getting the test person to verify the number of books read")
                    throw getPersonAfterUpdatingResult.error
                }
                log.info("Retrieved test person")
                
                const newPersonBooksRead = getPersonAfterUpdatingResult.value.getBooksRead()
                
                if(TEST_NUM_BOOKS_READ !== newPersonBooksRead){
                    throw new Error(`Expected number of book read ${TEST_NUM_BOOKS_READ} != actual number of books read ${newPersonBooksRead}`)
                }
            }finally{
                apiClientCloseFunction()
            }
        }finally{
            datastoreClientCloseFunction()
        }
    }finally{
        stopEnclaveFunction()
    }
});